### ISSUE TITLE
Problem with securing self-hosted Supabase

### ISSUE THREAD

**Issue by ymilhahn** [supabase/supabase#1583]
Hello everybody,

I would like to use Supabase for a project, but can't use the hosted service because of GDPR issues (missing order data processing, AWS as a sub contractor, ...). Therefore I tried the self-hosted version.

My setup is highly inspired by the Traefik-based solution from @firstdorsal (Thanks, Paul!)

# Bug report
## 1. /rest is reachable without kong (✔️ [solved by this commit](https://github.com/supabase/supabase/commit/2200e0406b2b29875d892a074d35e65a99ea0b07))

This is of course a security problem, because the authorization can easily be bypassed.
I suppose this is a specific problem of the Traefik solution as it is right now.
The [official docker solution](https://github.com/supabase/supabase/blob/master/docker/README.md) is providing everything (rest, auth, realtime) only behind kong. Is that right? So the Traefik-version should probably do the same? Right now I have to explicitly access the REST API via /kong/rest/v1, instead of /rest.

## 2. Row Level Security doesn't work

I've used RLS on hosted Supabase without any problems. Only the allowed data was returned.
However after exporting the schema and data from the hosted version and importing it into the self-hosted postgres. The RLS wasn't working anymore, despite that pgAdmin showed that RLS is active on that table.

### To Reproduce

1. Create Supabase project [with at least one table with RLS](https://supabase.io/docs/guides/auth#policy-examples)
2. Create a docker based self-hosted Supabase (I did it with Traefik, but that's probably not important)
3. [Regenerate the anon-key and service key like in these instructions](https://supabase.io/docs/guides/self-hosting#running-supabase)
4. Export the schema and data similar ["Migrate between projects"](https://supabase.io/docs/guides/database#migrating-between-projects), however, here I've migrated only public (schema and data) like recommended in the issue #1416.
I've search and replaced supabase_admin with postgres, because the self-hosted version doesn't have a supabase_admin.
5. Verify that RLS is active and and the Policies have also been imported (via pgAdmin)
6. Send a request:

docker-compose is showing those two lines on the request:
`curl --location --request GET 'https://supabase.my-server.com/kong/rest/v1/mytable?select=*' \
--header 'apikey: the-anon-key-generated-in-step-3'`

> supabase-rest | 172.22.0.4 - - [16/May/2021:14:12:21 +0000] "GET /mytable?select=* HTTP/1.1" 200 - "" "curl/7.76.1"
> supabase-kong | 172.22.0.2 - - [16/May/2021:14:12:21 +0000] "GET /rest/v1/mytable?select=* HTTP/1.1" 200 727 "-" "curl/7.76.1"
Note that rest is printing before kong (why?).

7. See that too much data is returned. An equivalent request to the hosted Supabase is returning the correct data.

**Probably kong is assigning the wrong role (not anon)?
Or the problem occurs because supabase_admin is not the owner anymore?**

bypassrls is only activated for the service_role ([like here](https://github.com/supabase/supabase/blob/master/docker/dockerfiles/postgres/00-initial-schema.sql#L15)).
I haven't modified auth-schema.sql nor the 00-initial-schema.sql.

### Expected behavior
Row level Security Policies are applied to the returned data by the REST API. 

## System information
- Server: Ubuntu 20.04
- Docker: 20.10.2

Postgres Roles:
![image](https://user-images.githubusercontent.com/2711970/118400590-8fc47800-b662-11eb-8be9-66de32e494b2.png)

**Comment by firstdorsal** [comment]
thank you! i'm glad it helped.
this is correct. i was not ware that kong handles the access control as i havent got to using supabase myself to be honest ^^
the last commit should fix this right?
https://github.com/supabase/supabase/pull/1562/commits/2200e0406b2b29875d892a074d35e65a99ea0b07

best, paul

**Comment by ymilhahn** [comment]
Thank you very much, Paul!
That definitely looks better and closes the biggest hole!

Regarding my second problem (Row Level Security not working via REST), I can now confirm that the RLS is definitely set up correctly in postgres.

`SET ROLE anon; SELECT * FROM mytable;` -> no data (like expected)
`RESET ROLE; SELECT * FROM mytable;` -> returns all entries (like expected)

So it really seems to be an related to kong not being able to assign the correct role (anon) to the provided key (anon-key).

**Comment by ymilhahn** [comment]
So, after some debugging I figured out that kong wasn't the problem here.

TL;DR: the **[PGRST_DB_ANON_ROLE](https://github.com/supabase/supabase/blob/master/docker/docker-compose.yml#L53)** variable in the for postgREST container was _postgres_, everything works fine with _anon_.

While debugging I found some very weird behavior:
- Users with the anon-key only were able to do everything (get, post, etc.)
- Logged in users with the anon-key and a Bearer JWT were correctly restricted by RLS.

Even after [revoking select rights for the _anon_ role](https://github.com/supabase/supabase/discussions/1493#discussioncomment-724842) the unauthenticated user was still able to do everything. So it was clear that _anon_ was never used, but _postgres_ was. Which isn't a surprise, since it's clearly stated in the [docker-compose.yml](https://github.com/supabase/supabase/blob/master/docker/docker-compose.yml#L53). Just took some time to find it...

After changing it to _anon_ everything worked properly. Just like in the hosted Supabase instance.

Is that by design or a mistake? @kiwicopple

**Comment by kiwicopple** [comment]
That looks like a bug @ymilhahn - well spotted. I believe when we first set this up it was using `postgres` as `anon` and we have since changed it. Would you mind making a PR to fix this one?

**Comment by ymilhahn** [comment]
Fixed by #1692.
